kalite/contentload/management/commands/channels/khan.py

```diff
@@ -79,39 +90,100 @@ def build_full_cache(items, id_key="id"):
     """
     for item in items:
         for attribute in item._API_attributes:
+            logging.info("Fetching item information for {id}, attribute: {attribute}".format(id=item.get(id_key, "Unknown"), attribute=attribute))
             try:
                 dummy_variable_to_force_fetch = item.__getattr__(attribute)
                 if isinstance(item[attribute], list):
                     for subitem in item[attribute]:
                         if isinstance(subitem, dict):
+                            try:
+                                subitem = json.loads(subitem.toJSON())
+                            except AttributeError:
+                                True
                             if subitem.has_key("kind"):
                                 subitem = whitewash_node_data(
                                     {key: value for key, value in subitem.items()})
                                 denorm_data(subitem)
                 elif isinstance(item[attribute], dict):
+                    try:
+                        item[attribute] = json.loads(item[attribute].toJSON())
+                    except AttributeError:
+                        True
                     if item[attribute].has_key("kind"):
                         item[attribute] = whitewash_node_data(
                             {key: value for key, value in item.attribute.items()})
                         denorm_data(item[attribute])
             except APIError as e:
                 del item[attribute]
+        try:
+            item = json.loads(item.toJSON())
+        except AttributeError:
+            True
+        for key in item.keys():
+            if hasattr(item[key], '__call__'):
+                logging.warn('{kind}: {item} had a function on key {key}'.format(kind=item.get('kind', ''), item=item.get('id', ''), key=key))
+                del item[key]
     return {item["id"]: whitewash_node_data(item) for item in items}
 
 def retrieve_API_data(channel=None):
     khan = Khan()
 
+    logging.info("Fetching Khan topic tree")
+
     topic_tree = khan.get_topic_tree()
 
+    logging.info("Fetching Khan exercises")
+
     exercises = khan.get_exercises()
 
+    logging.info("Fetching Khan videos")
+
     content = khan.get_videos()
 
+    # Hack to hardcode the mp4 format flag on Videos.
+    for con in content:
+        con["format"] = "mp4"
+
     assessment_items = []
 
-    # for exercise in exercises:
-    #     for assessment_item in exercise.all_assessment_items:
-    #         assessment_items.append(khan.get_assessment_item(assessment_item["id"]))
+    # Limit number of simultaneous requests
+    semaphore = threading.BoundedSemaphore(100)
+
+    def fetch_assessment_data(exercise):
```

Comment:
Will put a TODO (aron): refactor comment here, as another candidate for breaking up.